src/mbed_tools/build/_internal/config/assemble_build_config.py

```diff
@@ -24,8 +26,15 @@ def assemble_config(target_attributes: dict, mbed_program_directory: Path, mbed_
     Unfortunately, mbed_app.json may also contain filter labels to tell us which mbed libs we're depending on.
     This means we have to collect the filter labels from mbed_app.json before parsing any other config files.
     Then we parse all the required mbed_lib config and finally apply the app overrides.
+
+    Args:
+        target_attributes: Mapping of target specific config parameters.
+        search_paths: Iterable of paths to search for mbed_lib.json files.
+        mbed_app_file: The path to mbed_app.json. This can be None.
     """
-    mbed_lib_files = find_files("mbed_lib.json", mbed_program_directory)
+    mbed_lib_files = list(
+        set(itertools.chain.from_iterable(find_files("mbed_lib.json", path) for path in search_paths))
```

Comment:
What happens if multiple lib files are found? Using `set` would mean the order isn't guaranteed, so sometimes it might work and sometimes maybe not. I'd guess the user expectation is to use the deepest level mbed-os found. So, if the directory structure is `mbed-os/examples/some-example/mbed-os`, `mbed-os/examples/some-example/mbed-os` would be used and the top-level `mbed-os` would not be used.

Reply:
I'm not sure I understand what you mean. Multiple `mbed_lib.json` files are nearly always found. The intermediate cast to `set` removes any duplicate `mbed_lib.json` paths. It shouldn't affect how deeply we traverse into the directory tree. Maybe I'm misunderstanding the issue?

Anyway, the reasoning for the deduplication is: now that we always search `mbed_os.root` and `project.root` for all `mbed_lib.json` files, if `mbed_os.root` was a sub-directory of `project.root` we would search the `mbed-os` tree twice. Obviously this would cause many duplicate `mbed_lib.json` paths to end up in `mbed_lib_files`. If we had a directory structure like your example, I would expect the `mbed_lib.json` paths in the "top level" `mbed-os` to compare `neq` to the paths in the deepest level `mbed-os` so they would all be collected if both `mbed-os` folders were in either the `project.root` tree or the `mbed_os.root` tree.

Reply:
Sorry, I meant `mbed-os.lib`, which this line doesn't have anything to do with.